vnext/JSI/Shared/ChakraRuntime.cpp

```diff
@@ -607,284 +616,202 @@ facebook::jsi::Runtime::ScopeState *ChakraRuntime::pushScope() {
-    JsValueRef value;
-    JsGetUndefinedValue(&value);
-    return value;
-  } else if (valueIn.isNull()) {
-    JsValueRef value;
-    JsGetNullValue(&value);
-    return value;
-  } else if (valueIn.isBool()) {
-    JsValueRef value;
-    JsBoolToBoolean(valueIn.getBool(), &value);
-    return value;
-  } else if (valueIn.isNumber()) {
-    JsValueRef value;
-    JsDoubleToNumber(valueIn.getNumber(), &value);
-    return value;
-  } else if (valueIn.isString()) {
-    return stringRef(valueIn.getString(*this));
-  } else if (valueIn.isObject()) {
-    return objectRef(valueIn.getObject(*this));
-  } else {
-    // What are you?
-    abort();
-  }
-}
 
-ChakraRuntime::ChakraPropertyIdValue::~ChakraPropertyIdValue() {
-  JsRelease(m_propId, nullptr);
+  // Control flow should never reach here.
+  std::terminate();
 }
 
-void ChakraRuntime::ChakraPropertyIdValue::invalidate() {
-  delete this;
-}
+ChakraObjectRef ChakraRuntime::ToChakraObjectRef(
+    const facebook::jsi::Value &value) {
+  if (value.isUndefined()) {
+    JsValueRef ref;
+    ThrowUponJsError(JsGetUndefinedValue(&ref), "JsGetUndefinedValue");
+    return ChakraObjectRef(ref);
 
-ChakraRuntime::ChakraPropertyIdValue::ChakraPropertyIdValue(
-    JsPropertyIdRef propIdRef)
-    : m_propId(propIdRef) {
-  JsAddRef(propIdRef, nullptr);
-}
+  } else if (value.isNull()) {
+    JsValueRef ref;
+    ThrowUponJsError(JsGetNullValue(&ref), "JsGetNullValue");
+    return ChakraObjectRef(ref);
 
-ChakraRuntime::ChakraStringValue::ChakraStringValue(JsValueRef str)
-    : m_str(str) {
-  JsAddRef(str, nullptr);
-}
+  } else if (value.isBool()) {
+    JsValueRef ref;
+    ThrowUponJsError(JsBoolToBoolean(value.getBool(), &ref), "JsBoolToBoolean");
+    return ChakraObjectRef(ref);
 
-void ChakraRuntime::ChakraStringValue::invalidate() {
-  delete this;
-}
+  } else if (value.isNumber()) {
+    JsValueRef ref;
+    ThrowUponJsError(
+        JsDoubleToNumber(value.asNumber(), &ref), "JsDoubleToNumber");
+    return ChakraObjectRef(ref);
 
-ChakraRuntime::ChakraStringValue::~ChakraStringValue() {
-  JsRelease(m_str, nullptr);
-}
+  } else if (value.isSymbol()) {
+    return GetChakraObjectRef(value.asSymbol(*this));
 
-ChakraRuntime::ChakraObjectValue::ChakraObjectValue(JsValueRef obj)
-    : m_obj(obj) {
-  JsAddRef(m_obj, nullptr);
-}
+  } else if (value.isString()) {
+    return GetChakraObjectRef(value.asString(*this));
 
-void ChakraRuntime::ChakraObjectValue::invalidate() {
-  delete this;
-}
+  } else if (value.isObject()) {
+    return GetChakraObjectRef(value.asObject(*this));
 
-ChakraRuntime::ChakraObjectValue::~ChakraObjectValue() {
-  JsRelease(m_obj, nullptr);
+  } else {
+    // Control flow should never reach here.
+    std::terminate();
+  }
 }
 
-ChakraRuntime::ChakraWeakRefValue::ChakraWeakRefValue(JsWeakRef obj)
-    : m_obj(obj) {
-  JsAddRef(m_obj, nullptr);
-}
+std::vector<ChakraObjectRef> ChakraRuntime::ToChakraObjectRefs(
+    const facebook::jsi::Value *value,
```

Comment:
>const facebook::jsi::Value *value, [](start = 4, length = 34)

Since we control this code, we could just pass the safer std::vector through, or something like span<>

Reply:
The usage of Value* and and size is due the signatures of the jsi::call() and jsi::callAsConstructor(). Since we are not using C++ 20, std::span is not available. I can define a span-like struct here, but that seems a little overkill.

---
In reply to: [336277933](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336277933) [](ancestors = 336277933)